QBI Helper docs (#902)
spatialos/gdk-for-unity: docs/modules/qbi-helper/interest-template.md

<%(TOC)%>

# How to use InterestTemplate

<%(Callout message="

> Before reading this document, make sure have read:

Author: ```suggestion
Before reading this document, make sure you have read:
```

  * [Introduction to Query-based interest]({{urlRoot}}/modules/qbi-helper/intro-to-qbi)
")%>

> An [`InterestTemplate`]({{urlRoot}}/api/query-based-interest/interest-template) is effectively a wrapper around the `Interest` component, providing ways to more easily add, replace and clear queries from the underlying interest map.

Reviewer: ```
An [`InterestTemplate`]({{urlRoot}}/api/query-based-interest/interest-template) is effectively a wrapper around the `Interest` component, providing ways to more easily add, replace and clear queries from the underlying interest map.
```

to 

```
An [`InterestTemplate`]({{urlRoot}}/api/query-based-interest/interest-template) is a wrapper around the `Interest` component, providing intuitive methods to add, replace and clear queries from the underlying interest map.
```

> ## Initialise

Reviewer: Its kinda weird the heading is called `Initialise` but really you are Creating a template.

You initialise an empty [`InterestTemplate`]({{urlRoot}}/api/query-based-interest/interest-template) by calling `InterestTemplate.Create()`. This is useful when defining a set of interest queries for the first time, for example when defining Entity Templates.

```csharp
var basicInterestTemplate = InterestTemplate.Create();
```

You can also construct this from an existing [`InterestTemplate`]({{urlRoot}}/api/query-based-interest/interest-template) or set of interest queries. This creates a deep copy of the [`InterestTemplate`]({{urlRoot}}/api/query-based-interest/interest-template), to allow you to modify the queries without affecting the original set.

```csharp
var advancedInterestTemplate = InterestTemplate.Create(basicInterestTemplate);
```

## Modify

When modifying an [`InterestTemplate`]({{urlRoot}}/api/query-based-interest/interest-template), you must specify which particular component's queries you are modifying. You can either:

> * provide the component as a type argument

Reviewer: Type argument or generic argument? Not sure what the standard is

Author: If we were talking about it in the method, I'd call it a `generic type parameter` - so perhaps here we should cal it a `generic type argument`?

Author: calling it a `generic type parameter`

* pass the component ID as the first argument

For example, to add a query you might do one of the following:

```csharp
// Component identified by type argument
InterestTemplate.Create()
    .AddQueries<Position.Component>(query1, query2);

// Component identified by ID given as first argument
InterestTemplate.Create()
    .AddQueries(Position.ComponentId, query1, query2);
```

### Add

When adding queries, you can either provide the queries as parameters or an enumerable set:

```csharp
// Parameters
InterestTemplate.Create()
    .AddQueries<Position.Component>(query1, query2);

// Enumerable
var queryList = new List() { query1, query2 };
InterestTemplate.Create()
    .AddQueries<Position.Component>(queryList);
```

### Replace

> By doing this, you replace all the existing queries for an authoritative component with the new queries you pass in. In a similar way to adding queries, you must provide the queries as parameters or an enumerable set.

Reviewer: `By doing this, you replace` -> `This operation replaces`

Reviewer: `an authoritative component` -> `the given component id`

Reviewer: `you must provide` -> `you can provide`?

```csharp
// Parameters
InterestTemplate.Create(basicInterestTemplate)
    .ReplaceQueries<Position.Component>(query1);

// Enumerable
var queryList = new List() { query1 };
InterestTemplate.Create(basicInterestTemplate)
    .ReplaceQueries<Position.Component>(queryList);
```

> At least one query must be provided, otherwise the existing queries for a given authoritative component will not be removed.

Reviewer: This could be in a `>` block?

### Clear

You can either clear queries for a given authoritative component or clear _all_ queries in the [`InterestTemplate`]({{urlRoot}}/api/query-based-interest/interest-template).

```csharp
// Removes all queries added for the Position component
basicInterestTemplate.ClearQueries<Position.Component>();

// Removes all queries
basicInterestTemplate.ClearAllQueries();
```

> After adding, removing or modifying a set of queries, there are two ways to get `Interest` out of the [`InterestTemplate`]({{urlRoot}}/api/query-based-interest/interest-template).

Reviewer: This acts as a segue into the next section. But its feels like its part of `Clear`.

Suggest breaking this into two parts: `Modify` and `Apply Interest` or similar.

Then the segue can go under the `Apply Interest` header before you dive into `Snapshot` or `ComponentInterest`

Author: yup this is how I had it but I seem to have deleted the heading 😬 (adding it back)

```csharp
// Create an interest template
var interestTemplate = InterestTemplate.Create()
    .AddQueries<Position.Component>(query1, query2);

// Add the Interest component to an entity template
var entityTemplate = new EntityTemplate();
...
entityTemplate.AddComponent(interestTemplate.ToSnapshot(), WorkerUtils.UnityGameLogic);
```

### ComponentInterest

To update `Interest` at execution time, you need to modify the underlying Dictionary within the `Interest` component.

The [`InterestTemplate`]({{urlRoot}}/api/query-based-interest/interest-template) class provides the `AsComponentInterest` method to return this Dictionary, so that it can be used to update the `Interest` component.

For example, to update an entity's interest with a completely new set of queries in the MonoBehaviour workflow:

```csharp
// Create the new interest template
var newInterestTemplate = InterestTemplate.Create()
    .AddQueries<Position.Component>(query1, query2);

// Require the InterestWriter and update the ComponentInterest field

> InterestWriter.Data.ComponentInterest = newInterestTemplate.AsComponentInterest();

Reviewer: Don't use this as a demonstration. Its purely reference type semantics that cause the underlying data to change and the component isn't marked as dirty so it may never be replicated.

Instead use the `SendUpdate` method and create an Update object

<%(TOC)%>

# How to use InterestTemplate

<%(Callout message="
Before reading this document, make sure you have read:

  * [Introduction to Query-based interest]({{urlRoot}}/modules/qbi-helper/intro-to-qbi)
")%>

An [`InterestTemplate`]({{urlRoot}}/api/query-based-interest/interest-template) is a wrapper around the `Interest` component, providing intuitive methods to add, replace and clear queries from the underlying interest map.

## Create a template

> You create an empty [`InterestTemplate`]({{urlRoot}}/api/query-based-interest/interest-template) by calling `InterestTemplate.Create()`. This is useful when defining a set of interest queries for the first time, for example when defining Entity Templates.

Reviewer 2: pretty sure `Entity Templates` should not be captialized

```csharp
var basicInterestTemplate = InterestTemplate.Create();
```

> You can also construct this from an existing [`InterestTemplate`]({{urlRoot}}/api/query-based-interest/interest-template) or set of interest queries. This creates a deep copy of the [`InterestTemplate`]({{urlRoot}}/api/query-based-interest/interest-template), to allow you to modify the queries without affecting the original set.

Reviewer 2: ```suggestion
You can also construct this from an existing [`InterestTemplate`]({{urlRoot}}/api/query-based-interest/interest-template) or a set of interest queries. This creates a deep copy of the [`InterestTemplate`]({{urlRoot}}/api/query-based-interest/interest-template), to allow you to modify the queries without affecting the original set.
```

```csharp
var advancedInterestTemplate = InterestTemplate.Create(basicInterestTemplate);
```

## Modify a template

When modifying an [`InterestTemplate`]({{urlRoot}}/api/query-based-interest/interest-template), you must specify which particular component's queries you are modifying. You can either:

* provide the component as a type argument

> * pass the component ID as the first argument

Reviewer 2: ```suggestion
* provide the component ID as the first argument
```

to stay consistent

> For example, to add a query you might do one of the following:

Reviewer 2: can remove that sentence

```csharp
// Component identified by type argument
InterestTemplate.Create()
    .AddQueries<Position.Component>(query1, query2);

// Component identified by ID given as first argument
InterestTemplate.Create()
    .AddQueries(Position.ComponentId, query1, query2);
```

### Add

> When adding queries, you can either provide the queries as parameters or an enumerable set:

Reviewer 2: ```suggestion
When adding queries, you can either provide the queries as parameters or an enumerable set.
```

```csharp
// Parameters
InterestTemplate.Create()
    .AddQueries<Position.Component>(query1, query2);

// Enumerable
var queryList = new List() { query1, query2 };
InterestTemplate.Create()
    .AddQueries<Position.Component>(queryList);
```

### Replace

This operation replaces all the existing queries for the given component ID with the new queries you pass in. In a similar way to adding queries, you can provide the queries as parameters or an enumerable set.

```csharp
// Parameters
InterestTemplate.Create(basicInterestTemplate)
    .ReplaceQueries<Position.Component>(query1);

// Enumerable
var queryList = new List() { query1 };
InterestTemplate.Create(basicInterestTemplate)
    .ReplaceQueries<Position.Component>(queryList);
```

> > At least one query must be provided, otherwise the existing queries for a given authoritative component will not be removed.

Reviewer 2: ```suggestion
> At least one query must be provided, otherwise this operations does not remove the existing queries for a given authoritative component.
```

### Clear

You can either clear queries for a given authoritative component or clear _all_ queries in the [`InterestTemplate`]({{urlRoot}}/api/query-based-interest/interest-template).

```csharp
// Removes all queries added for the Position component
basicInterestTemplate.ClearQueries<Position.Component>();

// Removes all queries
basicInterestTemplate.ClearAllQueries();
```

## Get Interest from the template

After adding, removing or modifying a set of queries, there are two ways to get `Interest` out of the [`InterestTemplate`]({{urlRoot}}/api/query-based-interest/interest-template).

### Snapshot

To return a [`Snapshot`]({{urlRoot}}/api/core/snapshot) of the `Interest` component that can be used when defining entity templates, call `ToSnapshot()` on the [`InterestTemplate`]({{urlRoot}}/api/query-based-interest/interest-template).

```csharp
// Create an interest template
var interestTemplate = InterestTemplate.Create()
    .AddQueries<Position.Component>(query1, query2);

// Add the Interest component to an entity template
var entityTemplate = new EntityTemplate();

> ...

Reviewer 2: put this in a comment. We should try to provide valid code in our code snippets :)

entityTemplate.AddComponent(interestTemplate.ToSnapshot(), WorkerUtils.UnityGameLogic);
```

### ComponentInterest

To update `Interest` at execution time, you need to modify the underlying Dictionary within the `Interest` component.

> The [`InterestTemplate`]({{urlRoot}}/api/query-based-interest/interest-template) class provides the `AsComponentInterest` method to return this Dictionary, so that it can be used to update the `Interest` component.

Reviewer 2: ```suggestion
The [`InterestTemplate`]({{urlRoot}}/api/query-based-interest/interest-template) class provides the `AsComponentInterest` method to return this Dictionary. This can be used to update the `Interest` component.
```

> For example, to update an entity's interest with a completely new set of queries in the MonoBehaviour workflow:

Reviewer 2: remove this sentence
